File: Iverson.Clients/Python/iverson_client/aggregate.py
"""
Fluent aggregate builder that compiles to an ``AggregateRequest`` proto.

Unlike ``GroupByBuilder`` (one compound SELECT with all metrics as columns),
``AggregateRequest`` runs one bucketed/metric aggregation per ``AggregationSpec``
(TERMS/DATE_HISTOGRAM/RANGE buckets, or a single AVG/SUM/MIN/MAX/COUNT value),
with optional WHERE, HAVING, and JOIN — same shape as ``GroupByBuilder`` otherwise.

Usage:
    request = (
        aggregate("Article")
        .terms("category", "by_category", size=5)
        .avg("word_count", "avg_words")
        .where("is_published", SearchOperator.EQUALS, True)
        .build()
    )
"""
from __future__ import annotations

from typing import Optional, Sequence, Tuple

from google.protobuf.wrappers_pb2 import DoubleValue

from iverson_client.generated import object_search_pb2 as _pb
from iverson_client.search import _to_search_value


class AggregateBuilder:
    """Fluent DSL builder that compiles to an ``AggregateRequest`` proto message.

    Does not require a live server — ``build()`` simply returns the compiled
    proto. Instantiate via the module-level ``aggregate(type_name)`` factory.

    Note: ``AggregationSpec``'s ``group_by_fields``/``expression`` override
    fields (multi-key TERMS and raw-SQL-expression aggregations) are not
    covered by this builder.
    """

    def __init__(self, type_name: str) -> None:
        self._type_name = type_name
        self._aggregations: list[_pb.AggregationSpec] = []
        self._where: list[_pb.SearchClause] = []
        self._having: list[_pb.SearchClause] = []
        self._joins: list[_pb.JoinSpec] = []
        self._where_logic = _pb.AND
        self._having_logic = _pb.AND

    # ── WHERE filter (raw field strings, same value encoding as QueryBuilder) ──

    def where(self, field: str, op: int, value: object) -> "AggregateBuilder":
        """Add a WHERE (FILTER) clause."""
        self._where.append(_pb.SearchClause(
            property=field,
            operator=op,
            value=_to_search_value(value),
            clause_type=_pb.FILTER,
        ))
        return self

    def not_(self, field: str, op: int, value: object) -> "AggregateBuilder":
        """Add a MUST_NOT WHERE clause (excludes matches before aggregating)."""
        self._where.append(_pb.SearchClause(
            property=field,
            operator=op,
            value=_to_search_value(value),
            clause_type=_pb.MUST_NOT,
        ))
        return self

    def with_logic(self, logic: int) -> "AggregateBuilder":
        """Set the logic used to combine top-level WHERE clauses. Default: AND."""
        self._where_logic = logic
        return self

    # ── HAVING (references output alias names) ──────────────────────────────────

    def having(self, alias: str, op: int, value: object) -> "AggregateBuilder":
        """Add a HAVING clause. ``alias`` must match an aggregation's output name."""
        self._having.append(_pb.SearchClause(
            property=alias,
            operator=op,
            value=_to_search_value(value),
            clause_type=_pb.FILTER,
        ))
        return self

    def with_having_logic(self, logic: int) -> "AggregateBuilder":
        """Set the logic combining HAVING clauses. Default: AND."""
        self._having_logic = logic
        return self

    # ── JOIN ──────────────────────────────────────────────────────────────────

    def join(self, left_field: str, right_type: str, right_field: str,
              kind: int = _pb.JoinKind.INNER) -> "AggregateBuilder":
        """Add a join from this type to ``right_type`` on the given fields."""
        self._joins.append(_pb.JoinSpec(
            left_type=self._type_name,
            right_type=right_type,
            left_field=left_field,
            right_field=right_field,
            kind=kind,
        ))
        return self

    # ── Buckets ──────────────────────────────────────────────────────────────────

    def terms(self, field: str, name: str, size: int = 10) -> "AggregateBuilder":
        """Bucket by distinct values of ``field`` (up to ``size`` buckets)."""
        self._aggregations.append(_pb.AggregationSpec(
            name=name, type=_pb.TERMS, field=field, size=size,
        ))
        return self

    def date_histogram(self, field: str, name: str, calendar_interval: str,
                        time_zone: str = "") -> "AggregateBuilder":
        """Bucket a datetime ``field`` into calendar intervals (e.g. "day", "month")."""
        self._aggregations.append(_pb.AggregationSpec(
            name=name, type=_pb.DATE_HISTOGRAM, field=field,
            calendar_interval=calendar_interval, time_zone=time_zone,
        ))
        return self

    def range(self, field: str, name: str,
              buckets: Sequence[Tuple[Optional[str], Optional[float], Optional[float]]],
              ) -> "AggregateBuilder":
        """Bucket ``field`` into explicit ``(key, from, to)`` ranges. ``from``/``to`` of
        ``None`` means unbounded on that side."""
        range_buckets: list[_pb.RangeBucket] = []
        for key, lo, hi in buckets:
            kwargs: dict = {"key": key or ""}
            if lo is not None:
                kwargs["from"] = DoubleValue(value=lo)
            if hi is not None:
                kwargs["to"] = DoubleValue(value=hi)
            range_buckets.append(_pb.RangeBucket(**kwargs))
        self._aggregations.append(_pb.AggregationSpec(
            name=name, type=_pb.RANGE, field=field, range_buckets=range_buckets,
        ))
        return self

    # ── Metrics ──────────────────────────────────────────────────────────────────

    def avg(self, field: str, name: str) -> "AggregateBuilder":
        return self._add_metric(name, _pb.AVG, field)

    def sum(self, field: str, name: str) -> "AggregateBuilder":
        return self._add_metric(name, _pb.SUM, field)

    def min(self, field: str, name: str) -> "AggregateBuilder":
        return self._add_metric(name, _pb.MIN, field)

    def max(self, field: str, name: str) -> "AggregateBuilder":
        return self._add_metric(name, _pb.MAX, field)

    def count(self, field: str, name: str) -> "AggregateBuilder":
        return self._add_metric(name, _pb.COUNT, field)

    def count_all(self, name: str = "count") -> "AggregateBuilder":
        """COUNT(*) — leaves the aggregation's field empty."""
        return self._add_metric(name, _pb.COUNT, None)

    # ── Build ──────────────────────────────────────────────────────────────────

    def build(self, trace_id: str = "") -> _pb.AggregateRequest:
        """Compile to an ``AggregateRequest`` proto message.

        Raises:
            ValueError: If an aggregation ``name`` is duplicated.
        """
        names: set[str] = set()
        for a in self._aggregations:
            key = a.name.lower()
            if key in names:
                raise ValueError(f"Duplicate aggregation name '{a.name}'.")
            names.add(key)

        query = _pb.SearchQuery(
            clauses=self._where,
            logic=self._where_logic,
        )
        having = _pb.SearchQuery(
            clauses=self._having,
            logic=self._having_logic,
        )
        return _pb.AggregateRequest(
            type_name=self._type_name,
            query=query,
            aggregations=self._aggregations,
            having=having,
            joins=self._joins,
            trace_id=trace_id,
        )

    # ── Internal helpers ──────────────────────────────────────────────────────────

    def _add_metric(self, name: str, agg_type: int, field: Optional[str]) -> "AggregateBuilder":
        self._aggregations.append(_pb.AggregationSpec(
            name=name,
            type=agg_type,
            field=field or "",
        ))
        return self


def aggregate(type_name: str) -> AggregateBuilder:
    """Start an ``AggregateBuilder`` for the given entity type."""
    return AggregateBuilder(type_name)

File: Iverson.Clients/TypeScript/src/aggregate.ts
/**
 * Fluent aggregate builder that compiles to an AggregateRequest proto.
 *
 * Unlike GroupByBuilder (one compound SELECT with all metrics as columns),
 * AggregateRequest runs one bucketed/metric aggregation per AggregationSpec
 * (TERMS/DATE_HISTOGRAM/RANGE buckets, or a single AVG/SUM/MIN/MAX/COUNT value),
 * with optional WHERE, HAVING, and JOIN — same shape as GroupByBuilder otherwise.
 *
 * Note: AggregationSpec's groupByFields/expression override fields (multi-key
 * TERMS and raw-SQL-expression aggregations) are not covered by this builder.
 *
 * Usage:
 *   const req = aggregate('Article')
 *     .terms('category', 'byCategory', 5)
 *     .avg('wordCount', 'avgWords')
 *     .where('isPublished', SearchOperator.EQUALS, true)
 *     .build();
 */

import {
    AggregateRequest,
    AggregationSpec,
    AggregationType,
    JoinKind,
    JoinSpec,
    RangeBucket,
    SearchClause,
    SearchClauseType,
    SearchLogic,
    SearchOperator,
    SearchQuery,
} from '../generated/object_search.js';
import { toSearchValue } from './search.js';

export interface RangeBucketInput {
    key?: string;
    from?: number;
    to?: number;
}

// ── AggregateBuilder ──────────────────────────────────────────────────────────

/**
 * Fluent DSL builder that compiles to an AggregateRequest proto.
 */
export class AggregateBuilder {
    private readonly _typeName: string;
    private readonly _aggregations: AggregationSpec[] = [];
    private readonly _where: SearchClause[] = [];
    private readonly _having: SearchClause[] = [];
    private readonly _joins: JoinSpec[] = [];
    private _whereLogic: SearchLogic = SearchLogic.AND;
    private _havingLogic: SearchLogic = SearchLogic.AND;

    constructor(typeName: string) {
        this._typeName = typeName;
    }

    // ── Filtering ─────────────────────────────────────────────────────────────

    /** Add a WHERE clause (applied before aggregating). */
    where(field: string, op: SearchOperator, value: unknown): this {
        this._where.push({
            property: field,
            operator: op,
            value: toSearchValue(value),
            clauseType: SearchClauseType.FILTER,
        });
        return this;
    }

    /** Add a MUST_NOT WHERE clause (excludes matches before aggregating). */
    not(field: string, op: SearchOperator, value: unknown): this {
        this._where.push({
            property: field,
            operator: op,
            value: toSearchValue(value),
            clauseType: SearchClauseType.MUST_NOT,
        });
        return this;
    }

    /** Set the top-level WHERE clause logic (AND / OR). Default: AND. */
    withLogic(logic: SearchLogic): this {
        this._whereLogic = logic;
        return this;
    }

    /**
     * Add a HAVING clause (applied after aggregating). `alias` must be one of the
     * server's fixed output column aliases — `metric_val` for Avg/Sum/Min/Max/Count,
     * or `doc_count`/`bucket_key` for Terms/DateHistogram/Range — never the `name`
     * passed to a metric/bucket builder method. HAVING applies to every aggregation
     * in the request, not just this one.
     */
    having(alias: string, op: SearchOperator, value: unknown): this {
        this._having.push({
            property: alias,
            operator: op,
            value: toSearchValue(value),
            clauseType: SearchClauseType.FILTER,
        });
        return this;
    }

    /** Set the logic combining HAVING clauses. Default: AND. */
    withHavingLogic(logic: SearchLogic): this {
        this._havingLogic = logic;
        return this;
    }

    // ── Joins ─────────────────────────────────────────────────────────────────

    /** Add an inner/left/right join to another registered type. */
    join(leftField: string, rightType: string, rightField: string, kind: JoinKind = JoinKind.INNER): this {
        this._joins.push({
            leftType: this._typeName,
            rightType,
            leftField,
            rightField,
            kind,
        });
        return this;
    }

    // ── Aggregations — bucketing ────────────────────────────────────────────────

    /** Bucket by distinct values of `field` (up to `size` buckets). Default size: 10. */
    terms(field: string, name: string, size = 10): this {
        this._aggregations.push({
            name,
            type: AggregationType.TERMS,
            field,
            size,
            calendarInterval: '',
            timeZone: '',
            rangeBuckets: [],
            groupByFields: [],
            expression: '',
        });
        return this;
    }

    /** Bucket a datetime `field` into calendar intervals (e.g. "day", "month"). */
    dateHistogram(field: string, name: string, calendarInterval: string, timeZone = ''): this {
        this._aggregations.push({
            name,
            type: AggregationType.DATE_HISTOGRAM,
            field,
            size: 0,
            calendarInterval,
            timeZone,
            rangeBuckets: [],
            groupByFields: [],
            expression: '',
        });
        return this;
    }

    /** Bucket `field` into explicit ranges. A bucket's `from`/`to` left unset means unbounded on that side. */
    range(field: string, name: string, buckets: RangeBucketInput[]): this {
        const rangeBuckets: RangeBucket[] = buckets.map(b => ({
            key: b.key ?? '',
            from: b.from,
            to: b.to,
        }));
        this._aggregations.push({
            name,
            type: AggregationType.RANGE,
            field,
            size: 0,
            calendarInterval: '',
            timeZone: '',
            rangeBuckets,
            groupByFields: [],
            expression: '',
        });
        return this;
    }

    // ── Aggregations — metrics ─────────────────────────────────────────────────

    private _addMetric(name: string, type: AggregationType, field: string): this {
        this._aggregations.push({
            name,
            type,
            field,
            size: 0,
            calendarInterval: '',
            timeZone: '',
            rangeBuckets: [],
            groupByFields: [],
            expression: '',
        });
        return this;
    }

    /** AVG metric. */
    avg(field: string, name: string): this {
        return this._addMetric(name, AggregationType.AVG, field);
    }

    /** SUM metric. */
    sum(field: string, name: string): this {
        return this._addMetric(name, AggregationType.SUM, field);
    }

    /** MIN metric. */
    min(field: string, name: string): this {
        return this._addMetric(name, AggregationType.MIN, field);
    }

    /** MAX metric. */
    max(field: string, name: string): this {
        return this._addMetric(name, AggregationType.MAX, field);
    }

    /** COUNT metric on a specific field. */
    count(field: string, name: string): this {
        return this._addMetric(name, AggregationType.COUNT, field);
    }

    /** COUNT(*) metric — no field. Default alias: `count`. */
    countAll(name = 'count'): this {
        return this._addMetric(name, AggregationType.COUNT, '');
    }

    // ── Build ─────────────────────────────────────────────────────────────────

    /** Compile to an AggregateRequest proto message. */
    build(traceId = ''): AggregateRequest {
        const names = new Set<string>();
        for (const a of this._aggregations) {
            const key = a.name.toLowerCase();
            if (names.has(key)) throw new Error(`Duplicate aggregation name '${a.name}'.`);
            names.add(key);
        }

        const query: SearchQuery = {
            clauses: [...this._where],
            logic: this._whereLogic,
            sort: [],
        };
        const having: SearchQuery = {
            clauses: [...this._having],
            logic: this._havingLogic,
            sort: [],
        };
        return {
            typeName: this._typeName,
            query,
            aggregations: [...this._aggregations],
            having,
            joins: [...this._joins],
            traceId,
        };
    }
}

/** Start a fluent aggregate query for the given entity type. */
export function aggregate(typeName: string): AggregateBuilder {
    return new AggregateBuilder(typeName);
}
